### DISCUSSION TITLE
Colordistance sensor not resetting fast enough

### DISCUSSION THREAD

**Discussion by StPot3007** [pybricks/support#1747] [Q&A] [answered]
Hello community, I have a question if anyone has had issues with the colordistance sensor being slow in reading color changes.
I have a program where a train passes in front of the color sensor. The program stops the train when the correct color is detected by the sensor, but by the time the train comes to a halt the detected color is no longer aligned with the sensor, however when interrogated immediately after the stop the sensor still returns the color that was detected. Only by waiting up to 200ms does the sensor then detect the new color in front of it. Has anyone else come accross this behavior? I know that 200ms is not much and I have currently resolved my issue by having a loop after the original detection waiting up to 200ms and if no new color is returned I consider that the sensor is still aligned with the originally detected color. This seems to work but it is not ideal. Any ideas?

**Comment by laurensvalk** [comment, marked as answer]
The sensor is generally very fast. Can you share some example code where you have this issue?

**Reply by StPot3007** [reply]
Hi Laurensvalk,
Here is an extract from the code with comments. Sorry the code is a bit out of context, but I think it is understandable with the comments. Let me know if something is not clear.

```python
                if State=="Fwd 30" and instruction=="back 30 ok": #Enter the if statement with True on both conditions

                    instructionout="fwd 30"
                    print("Send " + instructionout)
                    data = ( str(broadcastChannel), instructionout )
                    hub.ble.broadcast(data)  #Send instruction to the train to move forward at speed 30

                    Totalwait=0
                    Maxwait=2000

                    while sensor.color() != Color.YELLOW:   #At this point the train is moving forward until the sensor detects color Yellow. It waits max 2000ms for this to happen

                        Totalwait=Totalwait+20

                        if Totalwait>Maxwait:
                            break               #Yellow detected faster than 2000ms
                        wait(20)

                    if Totalwait<2000:  #in this case the previous loop was interrupted before the 2000ms, so Yellow color was detected
                        if Totalwait==0:  #Yellow was immediatly detected (so sensor was immediately aligned with the color block)
                            print("on yellow")
                            State="stopfwd"
      
                        else:
                            print("Yellow " + str(Totalwait)) # in this case it took time Totalwait to detect the sensor
                            State="Back 30"
                    else:
                        print("No Yellow")  #In this case the sensor never detected the color
                        State="stop"


                if State=="stop" or State=="stopfwd" or State=="stopback":   #In the previous if statement the sensor detected the color within the 2000ms, so the State value becomes 'stopfwd', this will first send instruction to stop the train
                    print("Send 1st stop")
                    instructionout="stop" 
                    data = ( str(broadcastChannel), instructionout )
                    hub.ble.broadcast(data)

                    Totalwait=0
                    print(sensor.color())                    #At this point, the sensor still returns Yellow even though the sensor is no longer aligned with the yellow block
                    while sensor.color()==Color.YELLOW and Totalwait<200:    #The loop runs for max 200ms waiting for the sensor to return something else than yellow. 
                        #print(sensor.color())
                        #print(Totalwait)
                        Totalwait=Totalwait+20
                        wait(20)
                    print("Wait for sensor " + str(Totalwait))
                    if sensor.color()==Color.YELLOW: #At this point either yellow is still returned, meaning the sensor is still aligned with the color block, or Totalwait in the previous instruction is less than 200 (I have seen between 10ms and 150ms) then the color returned is something else than yellow.
                        instructionout="stop" 
                        State=""
                    else:
                        if State=="stopfwd":
                            instructionout="back 30"
                            State="Fwd 30"
                        else:
                            instructionout="fwd 30"
                            State="Back 30"

                    print("Send " + instructionout)
                    data = ( str(broadcastChannel), instructionout )
                    hub.ble.broadcast(data)
```

**Reply by laurensvalk** [reply]
When debugging an issue like this, I usually find it easier to take it one step at a time.

So perhaps you can try putting something together in a separate script that just measures color (and timing) and nothing else. This may help you see if it's the sensor or your script :)

**Reply by StPot3007** [reply]
I can do that and I'll send when done. Thanks for your time

**Reply by StPot3007** [reply]
Hi Laurens, I tried writing a small program to test the sensor reaction time (as shown below), but while testing I wonder if there really is a delay problem with the sensor reaction time. The way the program works is to show a yellow tile to the sensor and then quickly take it away. In this case I do get a 'wait' time where the sensor still sees the yellow tile, but I think it is just that I am not fast enough removing the tile from in front of the sensor (we are talking +- 80ms reaction time and sometimes less). This then made me think that maybe this is the same with the initial program where the train moves in front of the sensor, maybe I think the train has passed but in fact things go so quickly that the sensor is still seeing the tile for the time I believed the sensor was not 'resetting'. So unless you have an idea on how this could be tested in a better way I suggest to close this case. Thanks anyhow for your input.

```python
from pybricks.parameters import * # Color
# -----------------------------------------------
#  Set general values
# -----------------------------------------------
# -----------------------------------------------
#  Import classes and functions
# -----------------------------------------------


from pybricks.hubs import CityHub
from pybricks.pupdevices import DCMotor, Light, ColorDistanceSensor, Remote
from pybricks.parameters import Button, Color, Direction, Port, Side, Stop
from pybricks.robotics import DriveBase
from pybricks.tools import wait, StopWatch
from pybricks.iodevices import PUPDevice
from uerrno import ENODEV

hub = CityHub()
sensor = ColorDistanceSensor(Port.B)
x="N"

input("Ready ?")
Totalwait=0
Maxwait=2000
while sensor.color() != Color.YELLOW:

    Totalwait=Totalwait+20

    if Totalwait>Maxwait:
        break
    wait(20)

if Totalwait<2000:
    if Totalwait==0:
        print("on yellow")


    else:
        print("Yellow " + str(Totalwait))

else:
    print("No Yellow")

Totalwait=0
print(sensor.color())                    
while sensor.color()==Color.YELLOW and Totalwait<200:
    #print(sensor.color())
    #print(Totalwait)
    Totalwait=Totalwait+20
    wait(20)
print("Wait for sensor " + str(Totalwait))
```

**Reply by laurensvalk** [reply]
Sounds good!

> So unless you have an idea on how this could be tested in a better way I suggest to close this case.

If you want to be really systematic, you can test it against something you "know" the right value of. For example, you could make a motor push a few colored bricks underneath the sensor at a known speed. Then you can compare if the sensor is "slow" or actually giving the correct value :)

**Comment by StPot3007** [comment]
Thanks Laurent, this can be closed

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
